src/components/cylc/tree/index.js

```diff
@@ -1,6 +1,132 @@
 import { extractGroupState } from '@/utils/tasks'
 import { computePercentProgress } from '@/components/cylc'
 
+/**
+ * Create a workflow node. Uses the same properties (by reference) as the given workflow,
+ * only adding new properties such as type, children, etc.
+ *
+ * @param workflow {Object} workflow
+ * @return {{children: [], __type: string}}
+ */
+function createWorkflowNode (workflow) {
+  return {
+    __type: 'workflow',
+    ...workflow,
+    children: []
+  }
+}
+
+/**
+ * Create a cycle point node. Uses the family proxy property `cyclePoint`.
+ *
+ * @param familyProxy {Object} family proxy
+ * @return {{children: [], __type: string, name: *, id: *}}
+ */
+function createCyclePointNode (familyProxy) {
+  return {
+    __type: 'cyclepoint',
+    id: familyProxy.cyclePoint,
+    name: familyProxy.cyclePoint,
+    children: []
+  }
+}
+
+/**
+ * Create a family proxy node. Contains the same properties (by reference) as the given family,
+ * only adding new properties such as type, children, etc.
+ *
+ * @param familyProxy {Object} family proxy
+ * @return {{children: [], __type: string}}
+ */
+function createFamilyProxyNode (familyProxy) {
+  return {
+    __type: 'family',
+    ...familyProxy,
+    children: []
+  }
```

Comment:
@oliver-sanders I think this is what you meant with your previous example from the Graph component?

I tested and for the infinite tree we can use the same approach re-writing the state.

```js
return {
  __type: 'family',
  ...familyProxy,
  children: [],
  state: {
    open: true,
    size: 32,
    // etc, then
    state: familyProxy.state
  }
}
```

Reply:
And promise to keep the same for the infinite tree. Then there would still be the issue of how to handle the deltas :+1:

Reply:
Doesn't the splat operator duplicate the data?

The way the graph component does it is to put a reference to the object under it's own namespace:

https://github.com/cylc/cylc-ui/blob/697d383d281f6e158411c9128c38873fb18533c6/src/components/cylc/graph/Graph.vue#L253

So perhaps something like this:

```js
function createFamilyProxyNode (familyProxy) {
  return {
    __type: 'family',
    node: familyProxy,
    children: []
  }
```

Then you get a clear separation of Cylc data and Tree data.

```js
> treeItem.state
{
  open: true,
  size: 32
}
> treeItem.node.state
'succeeded'
```

Reply:
The `...` spread operator works similarly to `Object.assign`, so we could end up with copies. Changed to `node: { }`.

The `TreeItem.vue` funnily had a `node` prop already. So it became `node.node.state`... added 9 new commits to change the code and update required tests/other related code. Should we add another commit to rename it the `TreeItem` prop to `treeItem`? Or should we do that later @oliver-sanders ?

Quite happy with the result for now. It got simpler, and better organized this way, which should help with the infinite tree work later. (I've added a few TODO comments to help me remembering what goes where/why.)

Reply:
This looks good to me.